### ISSUE TITLE
Guided Root-on-ZFS Not Working With 1 Drive

### ISSUE THREAD

**Issue by seniorm0ment** [opnsense/installer#18]
Not sure if I'm in the right repo, if not please redirect me.
My system currently only has one drive in it.

When installing Opnsense, if I select to install using ZFS GPT/UEFI Hybrid mode, it asks me to select the drive(s) (offering the boot usb and the intended install ssd). I select stripe, it asks which drives, I select only the install ssd, it goes through and says it will format that, and proceeds and completes without error. On reboot, I pull the usb, it detects in bios and boots fine from the ssd working system.

However, if I choose other modes > auto zfs (guided root-on-zfs), default settings except enable encryption, encrypted swap, increase swap size, then under pool type/disks I choose stripe, and it asks which of the two drives, I select only the ssd, not the usb.
I then go and select proceed with install, and it says it will format that drive (it doesn't say anything about the USB). All goes fine, but when I reboot and pull the usb, my system refuses to boot from the ssd. It says no media found, insert drive and reboot. Now if I reboot but plug in the usb, it boots fine without issues, and asks for my encryption password I had set when installing. In case the issue wasn't clear, the USB should have nothing on it besides the livecd to install. The SSD is the drive I am installing to.

It worked no issues when selecting zfs gpt/uefi hybrid mode.
I wanted encryption enabled. I don't see why that would affect this, but I just can't boot without the usb when I install using zfs gpt/uefi hybrid.

**Comment by fichtner** [comment]
Hi @seniorm0ment,

Before getting further, can you clarify the following:

> When installing Opnsense, if I select to install using ZFS GPT/UEFI **Hybrid** mode [...] On reboot, I pull the usb, it detects in bios and **boots fine** from the ssd working system.

> I wanted encryption enabled. I don't see why that would affect this, but I just **can't boot without the usb** when I install using zfs gpt/uefi **hybrid**.


Cheers,
Franco

**Comment by seniorm0ment** [comment]
Thank you for your response, what exactly are you confused upon? For clarity, I only have one drive in this machine. OPNsense install/livecd is on a USB drive. My intentions are to install OPNsense from the USB, onto that one drive with ZFS+encryption.

When I boot up using the usb containing the Opnsense iso, it boots the opnsense cd, I login as instructed using user: installer. It offers the install methods. I tried two different ways.

The first way I tried was using the option that says Other modes > auto zfs (guided root-on-zfs). See my third paragraph on the original post for the walkthrough of the options I clicked. From what I can see, this way is the only way tk enable encryption on install with zfs.
After installing through, It asks to reboot, I let it. As it's rebooting I pull oit the livecd USB. At this point the only drive that is in the machine is the SSD (which I thought it installed opnsense to).
When it completes rebooting, it has trouble finding Opnsense on the ssd and won't boot. If I power down the machine, the turn it back on but this time with the livecd USB plugged back in, it works completely fine and boots. When it boots it asks for the encryption so I know it is the install and not just the livecd.
The issue is the USB should NOT be needed after I install, I ONLY selected the SSD under other modes > auto zfs > pool type/disks > stripe. It also said it would only format that one drive.

I had tried this install procesd twice, the second time powering off and unplugging the usb instead of just rebooting and pulling it out just to see. No cigar.
So basically the machine only boots when the usb is plugged in when it was installed using that method.

This however is NOT the case if I go back into the installer, and install to the SSD using the "ZFS GPT/UEFI Hybrid mode" option. I put stripe, only select the ssd, it installs, I reboot and pull out the livecd USB, and it boots into OPNsense and I have a working machine.
The issue is, this method doesn't offer those extended options (encryption, encrypted swap, etc). Only the method under "other modes > auto zfs" can I choose encryption. 
So that is why I am not satisfied with just using the ZFS GPT/UEFI Hybrid mode install.

**Comment by fichtner** [comment]
Well you gave a conflicting statement of the hybrid mode and I am trying to see if this was a typo or not.

The hybrid install is basically a copy of the "other modes" installation except that hybrid mode is maintained by us and the "other modes" installation is stock FreeBSD 12.1 code. Yes, it has extended options, but other drawbacks that hybrid mode does not have regarding UEFI/legacy bios interoperability.

That being said I would first make sure you can identify the custom option in "other modes" install that causes this. It's certainly one of those options. I may be able to help, but only if I can reproduce, but I don't want to test all and potentially come up empty.


Cheers,
Franco

**Comment by seniorm0ment** [comment]
Not a typo, I was going by how the menu lists and names them.
I did not know the hybrid install was a copy of the other modes section.
I guess I'm overall a bit confused on how I should be installing? The Hybrid mode option offers no options for encryption hence why I used the other modes option.

In other modes, the option I used was called "Auto ZFS (Guided Root-On-ZFS) and that opened the submenu to enable encryption and such. This was the mode that did not boot without the USB plugged.

>That being said I would first make sure you can identify the custom option in "other modes" install that causes this. It's certainly one of those options

Sorry I'm a bit confused as to what you want me to do when you say identify the custom optoon that causes this.
The option I selected was in custom modes > Auto ZFS (Guided Root-on-ZFS), whereas the option that had worked for me (without encryption) was the option on the main selection screen that stated "ZFS GPT/UEFI Hybrid mode"

**Comment by seniorm0ment** [comment]
Was this something that was being looked more into?
I hope I wasn't overly confusing, I was confusing myself a bit. Please let me know if you need more clarity, as I'm unsure if I properly understood what you wanted from me.
Or if you understood and just needed time.

**Comment by fichtner** [comment]
As I said, can you see which non-standard option you set in the FreeBSD based install routine that causes this behaviour? Or is this behaviour caused by the install routine even in default mode. As it stands there are too many variables to cover and the likelihood this can be fixed is with pushing the correct information to FreeBSD bug tracker eventually.

**Comment by seniorm0ment** [comment]
When you say non-standard option, you mean what options under the "other modes" menu?
As I originally stated, I enabled encryption, enabled encrypted swap, increased swap to 10gb, 

and then clicked the option pool type/disks, I chose stripe, and it asks which drives I want. I mark only the boot ssd, not the liveusb. It goes through and installs.

Are these not the options you are referring to by "non-standard" options?

**Comment by fichtner** [comment]
>  As I originally stated, I enabled encryption, enabled encrypted swap, increased swap to 10gb,

Sure, that's why I said one of these options may be causing the problem. Or it's the install routine itself. That the basic thing... we don't know.

**Comment by seniorm0ment** [comment]
Oh so you want me to test each one individually you're saying until I determine the problem? 
Will take a look and report back any findings.

**Comment by fichtner** [comment]
Yes that would be extremely helpful.

**Comment by GinjaChris** [comment]
I'm certainly not familiar with BSD install scripts but it sounds to me that if you select to encrypt the drive you then require the installation media to be present (on usb) otherwise the system won't boot.  I suspect therefore the encryption key is copied to usb, thus required at every reboot.

**Comment by fichtner** [comment]
22.1 comes out this week and the installer for ZFS features a number of fixes from FreeBSD so I am hoping this is resolved without knowing more about the conditions to reproduce which is why we can't leave this open as well.


Cheers,
Franco
